### DISCUSSION TITLE
Community Adapters

### DISCUSSION THREAD

**Discussion by KyleKing** [obi1kenobi/trustfall#418] [General]
It would be nice to have a list of Community Adapters as mentioned in #225!

I haven't used trustfall yet, but I would like to. I may have a use case where it would be nice to query across multiple log stores, such as looking for related logs by trace-id in AWS and Sumo Logic. If there were community adapters for each, they would likely wrap the network calls to return logs based on supported queries

Another use case might be to check for EOL software versions by comparing the versions in local toml/yaml/docker against a database of EOL code (https://endoflife.date) and/or release dates from the package manager (with rules for packages that are far behind recent releases)

**Comment by obi1kenobi** [comment]
Definitely! I'll look into it 👀 

I'd love to learn more about your use case as well, if you're open to sharing more details. What's generating the data? How are the queries and their results being used?

I don't believe there's an open-source adapter for AWS or Sumo Logic yet, but I'd be happy to help you write one if you'd like to build and open-source it! Or if this is a commercial use case, I'd be happy to write one and license it to your company.

**Reply by KyleKing** [reply]
My primary use case is after identifying an application log in Sumo Logic, I would want to find all related logs by `trace-id` in AWS CloudWatch across API Manager, EKS logs, micro-services that don't write to Sumo Logic, etc. The input would be the trace-id and some filter rules, and the output would be the set of JSONL logs across services. Beyond debugging, the exported logs could be useful for post-deployment quality assurance that requests are traceable.

However, I think AWS CloudWatch and Sumo Logic Adataptors aren't a good use case for Trustfall because Sumo and AWS already have respectively good and passable search functionality. The infrastructure changes to direct logs to Sumo or all to CloudWatch would be less than implementing and maintaining an adapter for these services.

I'll have to give more thought to other use cases

Our first use case might be to check dependency versions across `.tool-versions`, `Dockerfiles`, and GitHub Actions like you mentioned in the talk

**Reply by obi1kenobi** [reply]
Neat! If there's interest for a commercially-supported adapter, I'd be happy to chat about putting one together over whichever data source(s) you need.

> Sumo and AWS already have respectively good and passable search functionality

I assume you've already seen that Trustfall can wrap existing search functionality and extend it with additional expressiveness? For example, [this HackerNews playground query](https://play.predr.ag/hackernews#?f=1&q=IyBDcm9zcyBBUEkgcXVlcnkgKEFsZ29saWEgKyBGaXJlYmFzZSk6CiMgRmluZCBjb21tZW50cyBvbiBzdG9yaWVzIGFib3V0ICJvcGVuYWkuY29tIiB3aGVyZQojIHRoZSBjb21tZW50ZXIncyBiaW8gaGFzIGF0IGxlYXN0IG9uZSBHaXRIdWIgb3IgVHdpdHRlciBsaW5rCnF1ZXJ5IHsKICAjIFRoaXMgaGl0cyB0aGUgQWxnb2xpYSBzZWFyY2ggQVBJIGZvciBIYWNrZXJOZXdzLgogICMgVGhlIHN0b3JpZXMvY29tbWVudHMvdXNlcnMgZGF0YSBpcyBmcm9tIHRoZSBGaXJlYmFzZSBITiBBUEkuCiAgIyBUaGUgdHJhbnNpdGlvbiBpcyBzZWFtbGVzcyAtLSBpdCBpc24ndCB2aXNpYmxlIGZyb20gdGhlIHF1ZXJ5LgogIFNlYXJjaEJ5RGF0ZShxdWVyeTogIm9wZW5haS5jb20iKSB7CiAgICAuLi4gb24gU3RvcnkgewogICAgICAjIEFsbCBkYXRhIGZyb20gaGVyZSBvbndhcmQgaXMgZnJvbSB0aGUgRmlyZWJhc2UgQVBJLgogICAgICBzdG9yeVRpdGxlOiB0aXRsZSBAb3V0cHV0CiAgICAgIHN0b3J5TGluazogdXJsIEBvdXRwdXQKICAgICAgc3Rvcnk6IHN1Ym1pdHRlZFVybCBAb3V0cHV0CiAgICAgICAgICAgICAgICAgICAgICAgICAgQGZpbHRlcihvcDogInJlZ2V4IiwgdmFsdWU6IFsiJHNpdGVQYXR0ZXJuIl0pCgogICAgICBjb21tZW50IHsKICAgICAgICByZXBseSBAcmVjdXJzZShkZXB0aDogNSkgewogICAgICAgICAgY29tbWVudDogdGV4dFBsYWluIEBvdXRwdXQKCiAgICAgICAgICBieVVzZXIgewogICAgICAgICAgICBjb21tZW50ZXI6IGlkIEBvdXRwdXQKICAgICAgICAgICAgY29tbWVudGVyQmlvOiBhYm91dFBsYWluIEBvdXRwdXQKCiAgICAgICAgICAgICMgVGhlIHByb2ZpbGUgbXVzdCBoYXZlIGF0IGxlYXN0IG9uZQogICAgICAgICAgICAjIGxpbmsgdGhhdCBwb2ludHMgdG8gZWl0aGVyIEdpdEh1YiBvciBUd2l0dGVyLgogICAgICAgICAgICBsaW5rCiAgICAgICAgICAgICAgQGZvbGQKICAgICAgICAgICAgICBAdHJhbnNmb3JtKG9wOiAiY291bnQiKQogICAgICAgICAgICAgIEBmaWx0ZXIob3A6ICI%2BPSIsIHZhbHVlOiBbIiRtaW5Qcm9maWxlcyJdKQogICAgICAgICAgICB7CiAgICAgICAgICAgICAgY29tbWVudGVySURzOiB1cmwgQGZpbHRlcihvcDogInJlZ2V4IiwgdmFsdWU6IFsiJHNvY2lhbFBhdHRlcm4iXSkKICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICBAb3V0cHV0CiAgICAgICAgICAgIH0KICAgICAgICAgIH0KICAgICAgICB9CiAgICAgIH0KICAgIH0KICB9Cn0%3D&v=ewogICJzaXRlUGF0dGVybiI6ICJodHRwW3NdOi8vKFteLl0qXFwuKSpvcGVuYWkuY29tLy4qIiwKICAibWluUHJvZmlsZXMiOiAxLAogICJzb2NpYWxQYXR0ZXJuIjogIihnaXRodWJ8dHdpdHRlcilcXC5jb20vIgp9) first uses Algolia's search capability and then narrows down the results using the HackerNews Firebase REST API to answer "which Twitter or GitHub users comment on stories about OpenAI?"

So if Sumo or AWS search is decent but there's some missing bit of expressiveness that would be particularly useful, Trustfall could still be a fit.

**Comment by ZimboPro** [comment]
I have just published 3 adapters to crates.io, available here https://github.com/ZimboPro/trustfall-adapters. Adapters for OpenAPI, Terraform and a very simplified filesystem based off https://github.com/obi1kenobi/trustfall/discussions/225
